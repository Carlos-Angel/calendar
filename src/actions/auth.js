import { types } from '../types';
import { fetchWithoutToken, fetchWithToken } from '../helpers/fetch';
import Swa from 'sweetalert2';

export const startLogin = (email, password) => {
  return async (dispatch) => {
    const resp = await fetchWithoutToken(
      'auth/login',
      { email, password },
      'POST',
    );
    const body = await resp.json();
    if (body.ok) {
      localStorage.setItem('token', body.token);
      localStorage.setItem('token-init-date', new Date().getTime());
      dispatch(login(body.user));
    } else {
      Swa.fire('Error', body.msg, 'error');
    }
  };
};

export const startRegister = (email, password, name) => {
  return async (dispatch) => {
    const resp = await fetchWithoutToken(
      'auth/sign-in',
      { email, password, name },
      'POST',
    );
    const body = await resp.json();
    if (body.ok) {
      localStorage.setItem('token', body.token);
      localStorage.setItem('token-init-date', new Date().getTime());
      dispatch(login(body.user));
    } else {
      Swa.fire('Error', body.msg, 'error');
    }
  };
};

export const startChecking = () => {
  return async (dispatch) => {
    const resp = await fetchWithToken('auth/reset-token');
    const body = await resp.json();
    if (body.ok) {
      localStorage.setItem('token', body.token);
      localStorage.setItem('token-init-date', new Date().getTime());
      dispatch(login(body.user));
    } else {
      dispatch(checkingFinish());
    }
  };
};

export const startLogout = () => {
  return (dispatch) => {
    localStorage.clear();
    dispatch(logout());
  };
};

const login = (user) => ({
  type: types.authLogin,
  payload: user,
});

const checkingFinish = () => ({
  type: types.authCheckingFinish,
});

const logout = () => ({
  type: types.authStartLogout,
});
