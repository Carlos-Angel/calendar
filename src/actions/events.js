import { types } from '../types';
import { fetchWithToken } from '../helpers/fetch';
import { prepareEvents } from '../helpers/prepareEvents';

export const startEventAddNew = (event) => {
  return async (dispatch, getState) => {
    const { uid, name } = getState().auth;

    try {
      const resp = await fetchWithToken('event', event, 'POST');
      const body = await resp.json();
      if (body.ok) {
        const { id } = body.event;

        const user = {
          _id: uid,
          name,
        };

        dispatch(eventAddNew({ ...event, id, user }));
      }
    } catch (error) {
      console.log(error);
    }
  };
};

const eventAddNew = (event) => ({
  type: types.eventAddNew,
  payload: event,
});

export const eventSetActive = (event) => ({
  type: types.eventSetActive,
  payload: event,
});

export const eventClearActive = (event) => ({
  type: types.eventCleanActive,
});

export const eventUpdated = (event) => ({
  type: types.eventUpdated,
  payload: event,
});

export const eventDeleted = () => ({
  type: types.eventDeleted,
});

export const eventStartLoading = () => {
  return async (dispatch) => {
    try {
      const resp = await fetchWithToken('event');
      const body = await resp.json();
      if (body.ok) {
        const events = prepareEvents(body.events);
        dispatch(eventLoaded(events));
      }
    } catch (error) {
      console.log(error);
    }
  };
};

const eventLoaded = (events) => ({
  type: types.eventLoaded,
  payload: events,
});
