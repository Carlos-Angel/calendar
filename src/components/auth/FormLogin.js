import React from 'react';
import { useDispatch } from 'react-redux';
import { useForm } from '../../hooks/useForm';
import { startLogin } from '../../actions/auth';

export const FormLogin = () => {
  const dispatch = useDispatch();

  const [formValues, handleInputChange] = useForm({
    email: '',
    password: '',
  });

  const { email, password } = formValues;

  const onSubmit = (e) => {
    e.preventDefault();
    dispatch(startLogin(email, password));
  };

  return (
    <form onSubmit={onSubmit}>
      <div className='form-group'>
        <input
          type='text'
          className='form-control'
          placeholder='Correo'
          name='email'
          value={email}
          onChange={handleInputChange}
        />
      </div>
      <div className='form-group'>
        <input
          type='password'
          className='form-control'
          placeholder='ContraseÃ±a'
          name='password'
          value={password}
          onChange={handleInputChange}
        />
      </div>
      <div className='form-group'>
        <input type='submit' className='btnSubmit' value='Login' />
      </div>
    </form>
  );
};
