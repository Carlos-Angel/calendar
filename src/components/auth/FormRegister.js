import React from 'react';
import { useDispatch } from 'react-redux';
import { useForm } from '../../hooks/useForm';
import Swal from 'sweetalert2';
import { startRegister } from '../../actions/auth';

export const FormRegister = () => {
  const dispatch = useDispatch();

  const [formValues, handleInputChange] = useForm({
    name: '',
    email: '',
    password: '',
    confirmPassword: '',
  });

  const { name, email, password, confirmPassword } = formValues;

  const onSubmit = (e) => {
    e.preventDefault();
    if (password !== confirmPassword) {
      Swal.fire('Error', 'Las contraseñas no coinciden', 'error');
    } else {
      dispatch(startRegister(email, password, name));
    }
  };
  return (
    <form onSubmit={onSubmit}>
      <div className='form-group'>
        <input
          type='text'
          className='form-control'
          placeholder='Nombre'
          name='name'
          value={name}
          onChange={handleInputChange}
        />
      </div>
      <div className='form-group'>
        <input
          type='email'
          className='form-control'
          placeholder='Correo'
          name='email'
          value={email}
          onChange={handleInputChange}
        />
      </div>
      <div className='form-group'>
        <input
          type='password'
          className='form-control'
          placeholder='Contraseña'
          name='password'
          value={password}
          onChange={handleInputChange}
        />
      </div>

      <div className='form-group'>
        <input
          type='password'
          className='form-control'
          placeholder='Confirmar contraseña'
          name='confirmPassword'
          value={confirmPassword}
          onChange={handleInputChange}
        />
      </div>

      <div className='form-group'>
        <input type='submit' className='btnSubmit' value='Crear cuenta' />
      </div>
    </form>
  );
};
