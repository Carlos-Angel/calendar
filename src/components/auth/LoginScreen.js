import React from 'react';
import { FormLogin } from './FormLogin';
import { FormRegister } from './FormRegister';

import './login.css';

export const LoginScreen = () => {
  return (
    <div className='container login-container'>
      <div className='row'>
        <div className='col-md-6 login-form-1'>
          <h3>Ingreso</h3>
          <FormLogin />
        </div>

        <div className='col-md-6 login-form-2'>
          <h3>Registro</h3>
          <FormRegister />
        </div>
      </div>
    </div>
  );
};
