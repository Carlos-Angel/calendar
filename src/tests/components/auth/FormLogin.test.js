import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { FormLogin } from '../../../components/auth/FormLogin';
import {startLogin} from '../../../actions/auth';

import '@testing-library/jest-dom';

jest.mock('../../../actions/auth', () => ({
  startLogin: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initialState = {};
let store = mockStore(initialState);
store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={store}>
    <FormLogin />
  </Provider>,
);

describe('Pruebas en <FormLogin/>', () => {
  test('debe mostrarse correctamente', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('debe de llamar el dispatch del login', () => {
    wrapper.find('input[name="email"]').simulate('change', {
      target: {
        name: 'email',
        value: 'test@test.com',
      },
    });

    wrapper.find('input[name="password"]').simulate('change', {
      target: {
        name: 'password',
        value: 'password',
      },
    });

    wrapper.find('form').prop('onSubmit')({ preventDefault() {} });

    expect(startLogin).toHaveBeenCalledWith('test@test.com', 'password');
  });
});
