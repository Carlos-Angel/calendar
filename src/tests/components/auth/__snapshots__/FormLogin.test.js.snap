// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <FormLogin/> debe mostrarse correctamente 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="form-group"
  >
    <input
      className="form-control"
      name="email"
      onChange={[Function]}
      placeholder="Correo"
      type="text"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <input
      className="form-control"
      name="password"
      onChange={[Function]}
      placeholder="ContraseÃ±a"
      type="password"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <input
      className="btnSubmit"
      type="submit"
      value="Login"
    />
  </div>
</form>
`;
